docs/designs/termination.md

```diff
@@ -0,0 +1,130 @@
+
+We will additionally implement a Karpenter [Webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/) to validate node deletion requests and add finalizers to nodes that have been cleared for deletion. If the request will not violate a Node Disruption Budget (discussed below) and Karpenter is installed, the webhook will add the Karpenter finalizer to nodes and then allow the deletion request to go through, triggering the workflow. 
+
+### Eviction
+
+Node termination will only succeed when all pods are evicted. We use the [Kubernetes Eviction API](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#eviction-api) to handle eviction logic and [Pod Disruption Budget (PDB)](https://kubernetes.io/docs/tasks/run-application/configure-pdb/) violation errors.
+
+The Kubernetes Eviction API will only return a 200, 429, or 500 for each eviction call. A 200 indicates a successful request, requiring no error handling. A 429 indicates that this call would violate a PDB, and a 500 indicates a misconfiguration, such as multiple PDBs referring to the same pod. In the case of a 429 or 500, we will exponentially back off, sending a log in the Karpenter pod.  It’s possible (but very rare) that we violate a PDB with this API by sending multiple eviction requests to different master nodes simultaneously. If each eviction request would not violate a PDB individually but will when combined, it’s possible to violate a PDB. This is an API Server limitation and motivated the decision to evict pods serially per PDB to further reduce this chance (discussed below).
+
+While there is the [rare case](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#stuck-evictions) where stuck evictions require forceful termination, forcefully deleting a pod can have harsh repercussions in many cases. If evicting the pod would violate a PDB, it would essentially mean not supporting PDB. In addition, terminating pods that cannot afford downtime or [critical pods](https://kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/#marking-pod-as-critical) can have adverse effects on production workloads. Also, if a pod is owned by a StatefulSet, there is never a reason to forcibly delete, as mentioned [here](https://kubernetes.io/docs/tasks/run-application/force-delete-stateful-set-pod/). In the case that a user wants to forcefully terminate a pod, the user can terminate it with `kubectl delete pod <PODNAME> --grace-period=0 --force --namespace <NAMESPACE>` with the information in the Karpenter logs.
+
+After all pods evict, we will attempt to terminate the node. If the cloud-provider API returns an error, preventing instance termination, we will log the error in the Karpenter pod and node status then exponentially back off and retry.
```

Comment:
Thanks for the first one. The second I think is correct, but it can be misleading. We will log the error in the Karpenter pod, and also change the node status. I'll make changes to reflect that more accurately.